allennlp/state_machines/interactive_beam_search.py

```diff
@@ -0,0 +1,143 @@
+from collections import defaultdict
+from typing import Dict, List, Tuple
+
+import torch
+
+from allennlp.state_machines import util
+from allennlp.state_machines.beam_search import BeamSearch, StateType
+from allennlp.state_machines.transition_functions import TransitionFunction
+
+
+class InteractiveBeamSearch(BeamSearch[StateType]):
+    """
+    This class is designed to be a drop-in-replacement for `BeamSearch` that allows you to
+    specify the initial realization of the sequence. The idea is that in an interactive
+    setting you can force the beam search down certain paths, making for interesting
+    interactive demos.
+
+    Although the underlying BeamSearch infrastructure is batched, here we assume that the
+    batch has only a single element.
+
+    IMPORTANT: We assume that the ``TransitionFunction`` that you are using returns possible next
+    states in sorted order, so we do not do an additional sort inside of
+    ``InteractiveBeamSearch.search()``.  If you're implementing your own ``TransitionFunction``,
+    you must ensure that you've sorted the states that you return.
+
+    Parameters
+    ----------
+    beam_size : int
+        The beam size to use.
+    initial_sequence : ``torch.Tensor``, optional (default = None)
+        An ``(initial_sequence_length,)`` tensor containing the start of the sequence.
+    per_node_beam_size : ``int``, optional (default = beam_size)
+        The maximum number of candidates to consider per node, at each step in the search.
+        If not given, this just defaults to `beam_size`. Setting this parameter
+        to a number smaller than `beam_size` may give better results, as it can introduce
+        more diversity into the search. See Freitag and Al-Onaizan 2017,
+        "Beam Search Strategies for Neural Machine Translation".
+    """
+    def __init__(self,
+                 beam_size: int,
+                 initial_sequence: torch.Tensor = None,
+                 per_node_beam_size: int = None) -> None:
+        super().__init__(beam_size, per_node_beam_size)
+        if initial_sequence is not None:
+            # construct_prefix_tree wants a tensor of shape (batch_size, num_sequences, sequence_length)
+            # so we need to add the first two dimensions in. This returns a list, but we're assuming
+            # batch size 1, so we extract the first element.
+            self._allowed_transitions = util.construct_prefix_tree(initial_sequence.view(1, 1, -1))[0]
+        else:
+            self._allowed_transitions = {}
+
+        self.beams: List[List[Tuple[float, List[int]]]] = []
```

Comment:
I find "beams" here a little jarring, as the original meaning of "beam search" was that there was one beam containing multiple paths or hypotheses, and calling each path a "beam" is a bit odd to me.  But that usage is growing, so maybe I should just deal with it.

Reply:
in this case the usage is sort of right? because self.beams[0] is a "beam" (that is, a list of the scored action histories at time 0). I guess really this is "beam_snapshots" or something like that, but it's not just *a* beam.